pkg/store/bucket.go

```diff
@@ -788,43 +794,28 @@ type bucketIndexReader struct {
 	block *bucketBlock
 	dec   *index.DecoderV1
 
-	loadedPostings map[labels.Label]*lazyPostings
+	loadedPostings []*lazyPostings
 	loadedSeries   map[uint64][]byte
 }
 
 func newBucketIndexReader(ctx context.Context, block *bucketBlock) *bucketIndexReader {
 	r := &bucketIndexReader{
-		ctx:            ctx,
-		block:          block,
-		dec:            &index.DecoderV1{},
-		loadedPostings: map[labels.Label]*lazyPostings{},
-		loadedSeries:   map[uint64][]byte{},
+		ctx:   ctx,
+		block: block,
+		dec:   &index.DecoderV1{},
+
+		loadedSeries: map[uint64][]byte{},
 	}
 	r.dec.SetSymbolTable(r.block.symbols)
 	return r
 }
 
-func (r *bucketIndexReader) preloadPostings() error {
-	if len(r.loadedPostings) == 0 {
-		return nil
-	}
-
-	all := make([]*lazyPostings, 0, len(r.loadedPostings))
-	for _, p := range r.loadedPostings {
-		all = append(all, p)
-	}
-	sort.Slice(all, func(i, j int) bool {
-		return all[i].ptr.Start < all[j].ptr.Start
-	})
-	// TODO(fabxc): detect gaps and split up into multiple requests as we do for chunks.
-	start := all[0].ptr.Start
-	end := all[len(all)-1].ptr.End
-
-	b, err := r.block.readIndexRange(r.ctx, int64(start), int64(end-start))
+func (r *bucketIndexReader) preloadPostings(ctx context.Context, postings []*lazyPostings, start int64, length int64) error {
+	b, err := r.block.readIndexRange(ctx, int64(start), length)
 	if err != nil {
 		return errors.Wrap(err, "read postings range")
 	}
-	for _, p := range all {
+	for _, p := range postings {
```

Comment:
apparently on cluster I am getting this:
```
level=debug name=thanos-store ts=2017-12-08T02:03:16.821769008Z caller=bucket.go:395 msg="register postings" duration=31.6µs
panic: runtime error: slice bounds out of range
goroutine 2908 [running]:
github.com/improbable-eng/thanos/pkg/store.(*bucketIndexReader).preloadPostings(0xc4eba13940, 0x12c7820, 0xc4eba13a80, 0xc4ecb6a098, 0x1, 0x1, 0x1076320, 0x254, 0x456e60, 0xc4eb99ccc0)
	/home/bartek/go/src/github.com/improbable-eng/thanos/pkg/store/bucket.go:828 +0x24a
github.com/improbable-eng/thanos/pkg/store.(*bucketIndexReader).(github.com/improbable-eng/thanos/pkg/store.preloadPostings)-fm(0x12c7820, 0xc4eba13a80, 0xc4ecb6a098, 0x1, 0x1, 0x1076320, 0x254, 0xc4ec33d560, 0xffffffffffffffff)
	/home/bartek/go/src/github.com/improbable-eng/thanos/pkg/store/bucket.go:399 +0x7d
github.com/improbable-eng/thanos/pkg/store.preloadPostings.func2(0x4339b8, 0xdff320)
	/home/bartek/go/src/github.com/improbable-eng/thanos/pkg/store/bucket.go:883 +0xa2
github.com/improbable-eng/thanos/vendor/github.com/oklog/run.(*Group).Run.func1(0xc4ecb720c0, 0xc4ecb6c0f0, 0xc4ecb6a0d0)
	/home/bartek/go/src/github.com/improbable-eng/thanos/vendor/github.com/oklog/run/group.go:38 +0x27
created by github.com/improbable-eng/thanos/vendor/github.com/oklog/run.(*Group).Run
	/home/bartek/go/src/github.com/improbable-eng/thanos/vendor/github.com/oklog/run/group.go:37 +0xa8
```

Reply:
Are you sure it is this version running. The line `bucket.go:828` is a comment on this branch.

Reply:
Hm.. I can see  `bucket.go:828` as `_, l, err := r.dec.Postings(b[p.ptr.Start-start : p.ptr.End-start])`. I can repro it locally and fixed it now.